src/frontend/app/features/service-catalog/add-service-instance/create-service-instance-helper.service.ts

```diff
@@ -0,0 +1,272 @@
+import { GetServicePlanVisibilities } from '../../../store/actions/service-plan-visibility.actions';
+import { GetService } from '../../../store/actions/service.actions';
+import { GetSpace } from '../../../store/actions/space.actions';
+import { AppState } from '../../../store/app-state';
+import {
+  entityFactory,
+  organizationSchemaKey,
+  serviceBrokerSchemaKey,
+  servicePlanVisibilitySchemaKey,
+  serviceSchemaKey,
+  spaceSchemaKey,
+  spaceWithOrgKey,
+} from '../../../store/helpers/entity-factory';
+import { createEntityRelationKey, createEntityRelationPaginationKey } from '../../../store/helpers/entity-relations.types';
+import { getPaginationObservables } from '../../../store/reducers/pagination-reducer/pagination-reducer.helper';
+import {
+  selectCreateServiceInstance,
+  selectCreateServiceInstanceServicePlan,
+} from '../../../store/selectors/create-service-instance.selectors';
+import { APIResource } from '../../../store/types/api.types';
+import { CreateServiceInstanceState } from '../../../store/types/create-service-instance.types';
+import { CloudFoundryEndpointService } from '../../cloud-foundry/services/cloud-foundry-endpoint.service';
+import { fetchVisiblePlans, getSvcAvailability } from '../services-helper';
+import { ServicePlanAccessibility } from '../services.service';
+import { EntityService } from '../../../core/entity-service';
+
+
+@Injectable()
+export class CreateServiceInstanceHelperService {
+  serviceEntityService: EntityService<APIResource<IService>>;
+  // Is instance being created from the Marketplace
+  public marketPlaceMode = false;
+
+  // Internal property, clients should use `isInitialised()`
+  initialised$ = new BehaviorSubject(false);
+  cfGuid$ = new BehaviorSubject(null);
+  serviceGuid$ = new BehaviorSubject(null);
+  service$ = new BehaviorSubject(null);
+  serviceBroker$ = new BehaviorSubject(null);
+  servicePlanVisibilities$ = new BehaviorSubject(null);
+  constructor(
+    private store: Store<AppState>,
+    private activatedRoute: ActivatedRoute,
+    private entityServiceFactory: EntityServiceFactory,
+    private paginationMonitorFactory: PaginationMonitorFactory
+  ) {
+
+    const { serviceId, cfId } = activatedRoute.snapshot.params;
+    if (!!serviceId && !!cfId) {
+      this.marketPlaceMode = true;
+      this.cfGuid$.next(cfId);
+      this.serviceGuid$.next(serviceId);
+      this.initialised$.next(true);
+      this.initBaseObservables(serviceId, cfId);
+    }
+  }
+
+
+  initBaseObservables = (serviceId: string, cfGuid: string) => {
+    this.serviceEntityService = this.entityServiceFactory.create(
+      serviceSchemaKey,
+      entityFactory(serviceSchemaKey),
+      serviceId,
+      new GetService(serviceId, cfGuid),
+      true
+    );
+    this.serviceEntityService.waitForEntity$.pipe(
+      filter(o => !!o && !!o.entity),
+      map(o => o.entity),
+      publishReplay(1),
+      refCount(),
+      first(),
+      tap(svc => {
+        this.service$.next(svc);
```

Comment:
This isn't an observable because the service is initialised in a deferred manner. Stuff that depends on this breaks since `service$` will be undefined until the service has been selected by the user.

Reply:
And `initBaseObservables` can be called multiple times, since the user can go back to the previous step and change the selection to another service. This would redefine the observable, which didnt feel right to me.